privval/file_test.go

```diff
@@ -143,118 +167,158 @@ func TestUnmarshalValidatorKey(t *testing.T) {
 }
 
 func TestSignVote(t *testing.T) {
-	assert := assert.New(t)
-
-	privVal, _, _ := newTestFilePV(t)
-
-	randbytes := cmtrand.Bytes(tmhash.Size)
-	randbytes2 := cmtrand.Bytes(tmhash.Size)
-
-	block1 := types.BlockID{
-		Hash:          randbytes,
-		PartSetHeader: types.PartSetHeader{Total: 5, Hash: randbytes},
-	}
-	block2 := types.BlockID{
-		Hash:          randbytes2,
-		PartSetHeader: types.PartSetHeader{Total: 10, Hash: randbytes2},
+	for _, keyType := range kt.ListSupportedKeyTypes() {
+		t.Run(keyType, func(t *testing.T) {
+			assert := assert.New(t)
+			chainID := "mychainid" + keyType
+
+			keyGenF := func() (crypto.PrivKey, error) {
+				return kt.GenPrivKey(keyType)
+			}
+			privVal, _, _ := newTestFilePV(t, keyGenF)
+
+			randbytes := cmtrand.Bytes(tmhash.Size)
+			randbytes2 := cmtrand.Bytes(tmhash.Size)
+
+			block1 := types.BlockID{
+				Hash:          randbytes,
+				PartSetHeader: types.PartSetHeader{Total: 5, Hash: randbytes},
+			}
+			block2 := types.BlockID{
+				Hash:          randbytes2,
+				PartSetHeader: types.PartSetHeader{Total: 10, Hash: randbytes2},
+			}
+
+			height, round := int64(10), int32(1)
+			voteType := types.PrevoteType
+
+			// sign a vote for first time
+			vote := newVote(privVal.Key.Address, height, round, voteType, block1)
+			v := vote.ToProto()
+			err := privVal.SignVote(chainID, v, false)
+			require.NoError(t, err, "expected no error signing vote")
+			vote.Signature = v.Signature
+			err = vote.ValidateBasic()
+			require.NoError(t, err)
+
+			// Verify vote signature
+			pubKey, err := privVal.GetPubKey()
+			require.NoError(t, err)
+			err = vote.Verify(chainID, pubKey)
+			require.NoError(t, err)
+
+			// try to sign the same vote again; should be fine
+			err = privVal.SignVote(chainID, v, false)
+			require.NoError(t, err, "expected no error on signing same vote")
+
+			// now try some bad votes
+			cases := []*types.Vote{
+				newVote(privVal.Key.Address, height, round-1, voteType, block1),   // round regression
+				newVote(privVal.Key.Address, height-1, round, voteType, block1),   // height regression
+				newVote(privVal.Key.Address, height-2, round+4, voteType, block1), // height regression and different round
+				newVote(privVal.Key.Address, height, round, voteType, block2),     // different block
+			}
+
+			for _, c := range cases {
+				cpb := c.ToProto()
+				err = privVal.SignVote(chainID, cpb, false)
+				require.Error(t, err, "expected error on signing conflicting vote")
+			}
+
+			// try signing a vote with a different time stamp
+			sig := vote.Signature
+			vote.Signature = nil
+			vote.Timestamp = vote.Timestamp.Add(time.Duration(1000))
+			v2 := vote.ToProto()
+			err = privVal.SignVote(chainID, v2, false)
+			require.NoError(t, err)
+			assert.Equal(sig, v2.Signature)
```

Comment:
The changes here are fixing a bug that was present for a few years (in test code, not production)